### PR TITLE
Add prefixes to macros 

### PR DESCRIPTION
**Pull request by s-Nick**
# Description

This PR adds `ONEAPI_ONEMKL_` prefix to all the variables defined in `config.hpp.in` to be conformant with [ES.33 C++ Core Guidelines](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Res-MACROS).

Fixes # (GitHub issue)

This change was requested in https://github.com/oneapi-src/oneMKL/issues/554

# Checklist

## All Submissions

- [x] Do all unit tests pass locally? Attach a log.
[portblas_log.txt](https://github.com/user-attachments/files/17026039/portblas_log.txt)
[portfft_log.txt](https://github.com/user-attachments/files/17026041/portfft_log.txt)
[nvidia_all_domains_log.txt](https://github.com/user-attachments/files/17026043/nvidia_all_domains_log.txt)
[cpu_mkl_all_domains_log.txt](https://github.com/user-attachments/files/17026046/cpu_mkl_all_domains_log.txt)
[arc_mklgpu_all_domains_log.txt](https://github.com/user-attachments/files/17026048/arc_mklgpu_all_domains_log.txt)

- [x] Have you formatted the code using clang-format?

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -34,12 +34,13 @@ endif()
 # Build options
 option(BUILD_SHARED_LIBS "Build dynamic libraries" ON)
 
+if(BUILD_SHARED_LIBS)
+  set(ONEAPI_ONEMKL_BUILD_SHARED_LIBS True)
+endif()
```
Comment (Rbiessy) [inline CMakeLists.txt]: It'd be best to move it below the comment line 64 to explain why this variable is set.

File: CMakeLists.txt [hunk 2]
```diff
@@ -34,12 +34,13 @@ endif()
...
+
 ## Backends
 option(ENABLE_MKLCPU_BACKEND "Enable the Intel oneMKL CPU backend for supported interfaces" ON)
 option(ENABLE_MKLGPU_BACKEND "Enable the Intel oneMKL GPU backend for supported interfaces" ON)
-if(ENABLE_MKLCPU_BACKEND)
-  option(ENABLE_MKLCPU_THREAD_TBB "Enable the use of Intel TBB with the oneMKL CPU backend" ON)
-endif()
```
Comment (Rbiessy) [inline CMakeLists.txt]: This can be kept here to keep the options together

File: CMakeLists.txt [hunk 3]
```diff
@@ -60,6 +61,62 @@ option(ENABLE_CUFFT_BACKEND "Enable the cuFFT backend for the DFT interface" OFF
...
+# the corresponding generated macro. This is done to be conformant with ES.33
+# C++ Core Guidelines
+if(ENABLE_MKLCPU_BACKEND)
+  option(ENABLE_MKLCPU_THREAD_TBB "Enable the use of Intel TBB with the oneMKL CPU backend" ON)
+  set(ONEAPI_ONEMKL_ENABLE_MKLCPU_BACKEND True)
+endif()
+
+if(ENABLE_MKLGPU_BACKEND)
+  set(ONEAPI_ONEMKL_ENABLE_MKLGPU_BACKEND True)
+endif()
```
Comment (Rbiessy) [inline CMakeLists.txt]: You can simplify this in one line with `set(ONEAPI_ONEMKL_ENABLE_MKLGPU_BACKEND ${ENABLE_MKLGPU_BACKEND})`. According to [this ](https://cmake.org/cmake/help/latest/command/configure_file.html#transformations) `cmakedefine` will only set the C macro if the CMake variable evaluates to true.

Also if we go with this solution I would suggest to set the variables with the `ONEAPI_ONEMKL_` prefix in a different scope to make sure that we won't change one variable without changing the other. This can be done if we create a function for this block of code https://github.com/oneapi-src/oneMKL/blob/develop/src/CMakeLists.txt#L48, set the variables inside this function and call it.

File: CMakeLists.txt [hunk 4]
```diff
@@ -129,8 +186,10 @@ if(CMAKE_CXX_COMPILER OR NOT ONEMKL_SYCL_IMPLEMENTATION STREQUAL "dpc++")
     string(REPLACE "\\" "/" CMAKE_CXX_COMPILER ${CMAKE_CXX_COMPILER})
   endif()
 else()
-  if(ENABLE_CUBLAS_BACKEND OR ENABLE_CURAND_BACKEND OR ENABLE_CUSOLVER_BACKEND OR ENABLE_CUFFT_BACKEND
-    OR ENABLE_ROCBLAS_BACKEND OR ENABLE_ROCRAND_BACKEND OR ENABLE_ROCSOLVER_BACKEND OR ENABLE_ROCFFT_BACKEND)
+  if(ENABLE_CUBLAS_BACKEND OR ENABLE_CURAND_BACKEND OR
+     ENABLE_CUSOLVER_BACKEND OR ENABLE_CUFFT_BACKEND OR
+     ENABLE_ROCBLAS_BACKEND OR ENABLE_ROCRAND_BACKEND OR
+     ENABLE_ROCSOLVER_BACKEND OR ENABLE_ROCFFT_BACKEND)
```
Comment (Rbiessy) [inline CMakeLists.txt]: Can you revert the changes that just add noise? There are a few occurrences.

File: CMakeLists.txt [hunk 5]
```diff
@@ -294,8 +353,8 @@ else()
 endif()
 
 if(DEFINED REF_BLAS_ROOT)
-  find_file(REF_BLAS_LIBNAME NAMES blas.dll libblas.so HINTS ${REF_BLAS_ROOT} PATH_SUFFIXES lib lib64)
-  find_file(REF_CBLAS_LIBNAME NAMES cblas.dll libcblas.so HINTS ${REF_BLAS_ROOT} PATH_SUFFIXES lib lib64)
+  find_file(ONEAPI_ONEMKL_REF_BLAS_LIBNAME NAMES blas.dll libblas.so HINTS ${REF_BLAS_ROOT} PATH_SUFFIXES lib lib64)
+  find_file(ONEAPI_ONEMKL_REF_CBLAS_LIBNAME NAMES cblas.dll libcblas.so HINTS ${REF_BLAS_ROOT} PATH_SUFFIXES lib lib64)
```
Comment (Rbiessy) [inline CMakeLists.txt]: We don't need to rename these occurrences in CMake files. There are a few other occurrences.

File: CMakeLists.txt [hunk 6]
```diff
@@ -60,6 +61,62 @@ option(ENABLE_CUFFT_BACKEND "Enable the cuFFT backend for the DFT interface" OFF
...
+  set(ONEAPI_ONEMKL_ENABLE_CUFFT_BACKEND True)
+endif()
+
+if(ENABLE_ROCFFT_BACKEND)
+  set(ONEAPI_ONEMKL_ENABLE_ROCFFT_BACKEND True)
+endif()
+
+if(ENABLE_PORTFFT_BACKEND)
+  set(ONEAPI_ONEMKL_ENABLE_PORTFFT_BACKEND True)
+endif()
```
Comment (Rbiessy) [inline CMakeLists.txt]: I think we're missing the defines for `ONEAPI_ONEMKL_REF_BLAS_LIBNAME` and `ONEAPI_ONEMKL_REF_CBLAS_LIBNAME` here.

File: CMakeLists.txt (hunk 3 above)
Comment (Rbiessy) [reply CMakeLists.txt]: The whole function can live in `src/CMakeLiksts.txt:48` instead of the global CMake file IMO.

File: CMakeLists.txt (hunk 1 above)
Comment (s-Nick) [reply CMakeLists.txt]: Thank you for your comments @Rbiessy ,
I address this one in [3e73ce3](https://github.com/oneapi-src/oneMKL/pull/571/commits/3e73ce34b84b3cb0ef6477794995446d983cde5d)

File: CMakeLists.txt (hunk 2 above)
Comment (s-Nick) [reply CMakeLists.txt]: Moved back in [3e73ce3](https://github.com/oneapi-src/oneMKL/pull/571/commits/3e73ce34b84b3cb0ef6477794995446d983cde5d)

File: CMakeLists.txt (hunk 4 above)
Comment (s-Nick) [reply CMakeLists.txt]: yes, that was for my own readability. Reverted in [c1f8414](https://github.com/oneapi-src/oneMKL/pull/571/commits/c1f8414d1b426513b8ea7d45bf10b6358bb378b3)

File: CMakeLists.txt (hunk 5 above)
Comment (s-Nick) [reply CMakeLists.txt]: Following applies also to the other [comment](https://github.com/oneapi-src/oneMKL/pull/571#discussion_r1763051305)

I think that changes the name here is better. `find_file()` function creates the variable we pass as first argument, so I think it is better to define only one variable and then use it, rather than define one variable here and then specify what we really want in another place.
A possibility is to move those line inside `src/CMakeLists.txt` with the other variables defined for the header generation, since these two variables have the same usage.

File: CMakeLists.txt (hunk 3 above)
Comment (s-Nick) [reply CMakeLists.txt]: Thank you for your feedback @Rbiessy, I addressed changes in [3e73ce3](https://github.com/oneapi-src/oneMKL/pull/571/commits/3e73ce34b84b3cb0ef6477794995446d983cde5d) and [e8256f5](https://github.com/oneapi-src/oneMKL/pull/571/commits/e8256f503e1be61b68c6a3c97a73f3e9da405521)

File: CMakeLists.txt [hunk 7]
```diff
@@ -37,6 +37,7 @@ option(BUILD_SHARED_LIBS "Build dynamic libraries" ON)
 ## Backends
 option(ENABLE_MKLCPU_BACKEND "Enable the Intel oneMKL CPU backend for supported interfaces" ON)
 option(ENABLE_MKLGPU_BACKEND "Enable the Intel oneMKL GPU backend for supported interfaces" ON)
+
```
Comment (Rbiessy) [inline CMakeLists.txt]: The 2 new lines in this file are not needed IMO

File: deps/googletest/cmake/internal_utils.cmake [hunk 8]
```diff
@@ -26,7 +26,7 @@ macro(fix_default_compiler_settings_)
              CMAKE_C_FLAGS_MINSIZEREL CMAKE_C_FLAGS_RELWITHDEBINFO
              CMAKE_CXX_FLAGS CMAKE_CXX_FLAGS_DEBUG CMAKE_CXX_FLAGS_RELEASE
              CMAKE_CXX_FLAGS_MINSIZEREL CMAKE_CXX_FLAGS_RELWITHDEBINFO)
-      if (NOT BUILD_SHARED_LIBS AND NOT gtest_force_shared_crt)
+      if (NOT ONEAPI_ONEMKL_BUILD_SHARED_LIBS AND NOT gtest_force_shared_crt)
```
Comment (Rbiessy) [inline deps/googletest/cmake/internal_utils.cmake]: The changes in this file need to be reverted

File: src/CMakeLists.txt [hunk 9]
```diff
@@ -46,6 +40,26 @@ foreach(domain ${TARGET_DOMAINS})
 endforeach()
 
 # Generate header with enabled backends for testing
+
+function(generate_header_file)
```
Comment (Rbiessy) [inline src/CMakeLists.txt:44]: The lines inside the functions should be indented with 2 spaces, like the `if()`

File: src/CMakeLists.txt [hunk 10]
```diff
@@ -23,12 +23,6 @@ if(WIN32 AND BUILD_SHARED_LIBS)
   list(APPEND ONEMKL_BUILD_COPT "-Donemkl_EXPORTS")
 endif()
 
-# portBLAS backend variables must be accessible here to correctly
-# generate the config file.
-set(ENABLE_PORTBLAS_BACKEND_INTEL_CPU OFF CACHE INTERNAL "")
-set(ENABLE_PORTBLAS_BACKEND_INTEL_GPU OFF CACHE INTERNAL "")
-set(ENABLE_PORTBLAS_BACKEND_AMD_GPU OFF CACHE INTERNAL "")
-set(ENABLE_PORTBLAS_BACKEND_NVIDIA_GPU OFF CACHE INTERNAL "")
```
Comment (Rbiessy) [inline src/CMakeLists.txt:31]: Are you sure we can remove these lines? That seemed needed at the time. Is it working with portBLAS disabled and the minimum CMake version supported 3.13?

File: CMakeLists.txt (hunk 7 above)
Comment (s-Nick) [reply CMakeLists.txt]: Fixed in [9ae43f0](https://github.com/oneapi-src/oneMKL/pull/571/commits/9ae43f0036d3bef1060383a467720988576fa5d6)

File: deps/googletest/cmake/internal_utils.cmake (hunk 8 above)
Comment (s-Nick) [reply deps/googletest/cmake/internal_utils.cmake]: fixed in [9ae43f0](https://github.com/oneapi-src/oneMKL/pull/571/commits/9ae43f0036d3bef1060383a467720988576fa5d6)

File: src/CMakeLists.txt (hunk 9 above)
Comment (s-Nick) [reply src/CMakeLists.txt:44]: fixed in [9ae43f0](https://github.com/oneapi-src/oneMKL/pull/571/commits/9ae43f0036d3bef1060383a467720988576fa5d6)

File: src/CMakeLists.txt (hunk 10 above)
Comment (s-Nick) [reply src/CMakeLists.txt:31]: yes I double checked with different version of cmake and all backends works correctly. Using cmake 3.13 I found other issues in `develop` branch. I'll open an issue for those soon.

**Review by Rbiessy — APPROVED** [review]
LGTM!

File: CMakeLists.txt [hunk 11]
```diff
@@ -294,8 +294,8 @@ else()
 endif()
 
 if(DEFINED REF_BLAS_ROOT)
-  find_file(REF_BLAS_LIBNAME NAMES blas.dll libblas.so HINTS ${REF_BLAS_ROOT} PATH_SUFFIXES lib lib64)
-  find_file(REF_CBLAS_LIBNAME NAMES cblas.dll libcblas.so HINTS ${REF_BLAS_ROOT} PATH_SUFFIXES lib lib64)
+  find_file(ONEAPI_ONEMKL_REF_BLAS_LIBNAME NAMES blas.dll libblas.so HINTS ${REF_BLAS_ROOT} PATH_SUFFIXES lib lib64)
```
Comment (lhuot) [inline CMakeLists.txt]: Overall the change looks okay to me, my only concern is about the macro prefix choice. ONEAPI_ONEMKL is a little bit redundant and heavy in my opinion, could we consider using only ONEMKL or follow the namespace with ONEAPI_MKL (this one would have my preference)?

Comment (Rbiessy) [reply CMakeLists.txt]: I'm not sure we could use `ONEAPI_MKL` as I vaguely remember we had to remove occurrences of "MKL" in the past. The exceptions are the name of the backends MKLCPU, MKLGPU and the mkl namespace.
I don't mind too much between `ONEAPI_ONEMKL` or `ONEMKL`.

Comment (s-Nick) [reply CMakeLists.txt]: Thank you for your review @lhuot. 
I agree it is quite verbose. In my opinion, the best option is to use `ONEMKL` for now, so that it can be renamed later in `ONEMATH`. What do you think?

Comment (lhuot) [reply CMakeLists.txt]: That's fine by me. @mkrainiuk any concern with the macro prefix that should be used?

Comment (mkrainiuk) [reply CMakeLists.txt]: I also support shorter name if possible (longer macro name increase the chance of a typo), `ONEMKL` looks great to me while we are migrating to oneMath, but does it make sense to start changing it to `ONEMATH` already to reduce users efforts for migrating to new macro names in their build configurations?

File: include/oneapi/mkl/detail/export.hpp [hunk 12]
```diff
@@ -22,7 +22,7 @@
 
 #include "oneapi/mkl/detail/config.hpp"
 
-#if !defined(BUILD_SHARED_LIBS) || !defined(_WIN64)
```
Comment (mkrainiuk) [inline include/oneapi/mkl/detail/export.hpp:25]: We rely here on the [standard CMake variable](https://cmake.org/cmake/help/latest/variable/BUILD_SHARED_LIBS.html) I don't think we need to introduce something new for this project.

File: src/config.hpp.in [hunk 13]
```diff
@@ -20,25 +20,25 @@
...
+#cmakedefine ONEAPI_ONEMKL_ENABLE_PORTBLAS_BACKEND_INTEL_GPU
+#cmakedefine ONEAPI_ONEMKL_ENABLE_PORTBLAS_BACKEND_NVIDIA_GPU
+#cmakedefine ONEAPI_ONEMKL_ENABLE_PORTFFT_BACKEND
+#cmakedefine ONEAPI_ONEMKL_ENABLE_ROCBLAS_BACKEND
+#cmakedefine ONEAPI_ONEMKL_ENABLE_ROCFFT_BACKEND
+#cmakedefine ONEAPI_ONEMKL_ENABLE_ROCRAND_BACKEND
+#cmakedefine ONEAPI_ONEMKL_ENABLE_ROCSOLVER_BACKEND
+#cmakedefine ONEAPI_ONEMKL_BUILD_SHARED_LIBS
+#cmakedefine ONEAPI_ONEMKL_REF_BLAS_LIBNAME "@ONEAPI_ONEMKL_REF_BLAS_LIBNAME@"
+#cmakedefine ONEAPI_ONEMKL_REF_CBLAS_LIBNAME "@ONEAPI_ONEMKL_REF_CBLAS_LIBNAME@"
```
Comment (mkrainiuk) [inline src/config.hpp.in]: These two can be removed now

File: CMakeLists.txt (hunk 11 above)
Comment (s-Nick) [reply CMakeLists.txt]: thank you for the feedback @mkrainiuk,
I don't like the idea of having these variables already renamed while everything else isn't. I know @Rbiessy  is working on it, but I don't want to have this period with different names in between his renaming patch and this one. Moreover, these variables change should not affect users but only internal headers. So I would rather use `ONEMKL` for now.

File: src/CMakeLists.txt [hunk 14]
```diff
@@ -46,15 +40,38 @@ foreach(domain ${TARGET_DOMAINS})
...
+  set(ONEAPI_ONEMKL_ENABLE_ROCBLAS_BACKEND ${ENABLE_ROCBLAS_BACKEND})
+  set(ONEAPI_ONEMKL_ENABLE_NETLIB_BACKEND ${ENABLE_NETLIB_BACKEND})
+  set(ONEAPI_ONEMKL_ENABLE_PORTBLAS_BACKEND ${ENABLE_PORTBLAS_BACKEND})
+  set(ONEAPI_ONEMKL_ENABLE_CURAND_BACKEND ${ENABLE_CURAND_BACKEND})
+  set(ONEAPI_ONEMKL_ENABLE_ROCRAND_BACKEND ${ENABLE_ROCRAND_BACKEND})
+  set(ONEAPI_ONEMKL_ENABLE_CUSOLVER_BACKEND ${ENABLE_CUSOLVER_BACKEND})
+  set(ONEAPI_ONEMKL_ENABLE_ROCSOLVER_BACKEND ${ENABLE_ROCSOLVER_BACKEND})
+  set(ONEAPI_ONEMKL_ENABLE_CUFFT_BACKEND ${ENABLE_CUFFT_BACKEND})
+  set(ONEAPI_ONEMKL_ENABLE_ROCFFT_BACKEND ${ENABLE_ROCFFT_BACKEND})
+  set(ONEAPI_ONEMKL_ENABLE_PORTFFT_BACKEND ${ENABLE_PORTFFT_BACKEND})
```
Comment (mkrainiuk) [inline src/CMakeLists.txt]: Does it make sense to mention this renaming from user input to the internal name in the documentation? So users will be aware what actual variables they need to check for debug purposes?

File: CMakeLists.txt (hunk 11 above)
Comment (mkrainiuk) [reply CMakeLists.txt]: I see, it's fine to me as long as it doesn't affect the users.

File: src/config.hpp.in (hunk 13 above)
Comment (s-Nick) [reply src/config.hpp.in]: Addressed while rebasing to include this change

File: src/CMakeLists.txt (hunk 14 above)
Comment (s-Nick) [reply src/CMakeLists.txt]: As mentioned in a [comment above](https://github.com/oneapi-src/oneMKL/pull/571#discussion_r1793768103), these definitions should not be used by users and they are internal.

File: include/oneapi/mkl/detail/export.hpp (hunk 12 above)
Comment (s-Nick) [reply include/oneapi/mkl/detail/export.hpp:25]: Thank you for your feedback @mkrainiuk 
`BUILD_SHARED_LIBS` is a standard variable in CMake, but, in this case, it was converted to be used inside C++ project. I don't think users should rely upon any macro defined in oneMath a part from the version macro.

File: CMakeLists.txt (hunk 11 above)
Comment (s-Nick) [reply CMakeLists.txt]: I changed names in [2a739a8](https://github.com/oneapi-src/oneMKL/pull/571/commits/2a739a8ff9ca5d965832b4146c52a06be6454d65)

File: include/oneapi/mkl/detail/export.hpp (hunk 12 above)
Comment (s-Nick) [reply include/oneapi/mkl/detail/export.hpp:25]: To clarify my comment, I believe users should avoid relying on any macros defined within this library for internal usage. While `BUILD_SHARED_LIBS` is a standard option in CMake, converting it into a C++ macro can lead to issues if users wish to use this macro in their own projects. Additionally, it’s not a C++ standard macro. Therefore, I think it would be safer to make this macro project-specific by adding the prefix, as it has been done with other macros in this PR

**Review by mkrainiuk — APPROVED** [review]
